### PR TITLE
Spring boot security

### PR DESCRIPTION
**Pull request by TetianaDenchyk**

### CODE REVIEW

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml
Comment (Alexsandr-Yablunovskyi) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml:1]: You are creating a user table but gives id: create-books-table and tableName: "books" and file name "01-create-books-table.yaml. And in the next file "02-create-users-table.yaml" you are creating a book table but table, id and file has a name "users".
I guess it's better to coordinate the name of the file and the logic written in it

**Review by Alexsandr-Yablunovskyi — APPROVED** [review]
Great job!!
Just a little comment about file naming in the "changes" directory.

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml
Comment (TetianaDenchyk) [reply src/main/resources/db/changelog/changes/01-create-books-table.yaml:1]: Thank you for noticing that! I will fix

File: pom.xml [hunk 1]
```diff
@@ -79,6 +80,30 @@
...
+		</dependency>
+		<dependency>
+			<groupId>io.jsonwebtoken</groupId>
+			<artifactId>jjwt-impl</artifactId>
+			<version>${jjwt.version}</version>
+			<scope>runtime</scope>
+		</dependency>
+
+
+
```
Comment (liliia-ponomarenko) [inline pom.xml]:
```suggestion
```

File: src/main/java/com/bookshop/annotation/FieldMatch.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.bookshop.annotation;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/annotation/FieldMatch.java]: It’s difficult to search that your annotations and validators are placed in different packages. better move it to the validator or make this package as subpackage of the validator

File: src/main/java/com/bookshop/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                .build();
+    }
+
+    @Bean
+    public AuthenticationManager authenticationManager(
+            AuthenticationConfiguration authenticationConfiguration
+    ) throws Exception {
+        return authenticationConfiguration.getAuthenticationManager();
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/config/SecurityConfig.java]:
```suggestion
```

File: src/main/java/com/bookshop/controller/BookController.java [hunk 4]
```diff
@@ -33,12 +34,14 @@ public class BookController {
     @GetMapping
     @Operation(summary = "Get all books by page",
             description = "Get a list of all books at x page with y size")
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/controller/BookController.java]: you can use hasAnyRole('USER', 'ADMIN')

File: src/main/java/com/bookshop/dto/user/UserRegistrationRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import jakarta.validation.constraints.Size;
+
+@FieldMatch
+public record UserRegistrationRequestDto(@NotBlank @Email String email,
+                                         @NotBlank @Size(min = 7, max = 20) String password,
+                                         @NotBlank @Size(min = 7, max = 20) String repeatPassword,
+                                         @NotBlank String firstName,
+                                         @NotBlank String lastName,
+                                         String shippingAddress) {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/dto/user/UserRegistrationRequestDto.java]:
```suggestion
```

File: src/main/java/com/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -29,6 +29,12 @@ public ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundExcept
                 .body(ex.getMessage());
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException ex) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/exception/CustomGlobalExceptionHandler.java]: CONFLICT more appropriate

File: src/main/java/com/bookshop/security/CustomUserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import com.bookshop.exception.EntityNotFoundException;
+import com.bookshop.repository.UserRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/security/CustomUserDetailsService.java]:
```suggestion
@Service 
```

File: src/main/java/com/bookshop/security/JwtAuthenticationFilter.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String AUTHORIZATION_HEADER = "Authorization";
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/security/JwtAuthenticationFilter.java]: You can use HttpHeader.AUTHORIZATION instead

File: src/main/java/com/bookshop/security/JwtUtil.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    private <T> T getClaimFromToken(String token, Function<Claims, T> claimsResolver) {
+        final Claims claims = Jwts.parser()
+                .verifyWith(secret)
+                .build()
+                .parseSignedClaims(token)
+                .getPayload();
+        return claimsResolver.apply(claims);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/security/JwtUtil.java]:
```suggestion
```

File: src/main/java/com/bookshop/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+    private final RoleRepository roleRepository;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.email()).isPresent()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/service/impl/UserServiceImpl.java]: better use existsByEmail from JPA

File: src/main/java/com/bookshop/validator/IsbnValidator.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package com.bookshop.validator;
+
+import com.bookshop.annotation.Isbn;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class IsbnValidator implements ConstraintValidator<Isbn, String> {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookshop/validator/IsbnValidator.java]:
```suggestion
```

File: src/main/resources/application.properties [hunk 12]
```diff
@@ -3,8 +3,11 @@ spring.datasource.url=jdbc:mysql://localhost:3306/online_bookshop_project
 spring.datasource.username=root
 spring.datasource.password=Great77prog.
 
-spring.jpa.hibernate.ddl-auto=validate
+spring.jpa.hibernate.ddl-auto=create-drop
```
Comment (liliia-ponomarenko) [inline src/main/resources/application.properties]: Why do you change it?

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 13]
```diff
@@ -14,28 +14,28 @@ databaseChangeLog:
...
                     nullable: false
               - column:
-                  name: isbn
+                  name: firstName
                   type: varchar(255)
                   constraints:
                     nullable: false
-                    unique: true
               - column:
-                  name: price
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]: Why do you change create-books-table and add columns for the user?

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+                  constraints:
+                    nullable: false
+              - column:
+                  name: last_name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(255)
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml:39]: make sure that it does not cause SQL exception (you should validate RequestDto)

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+              - column:
+                  name: is_deleted
+                  type: bit
+                  defaultValue: 'false'
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]: add one empty line to the end of the file (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/bookshop/service/impl/UserServiceImpl.java (hunk 10 above)
Comment (TetianaDenchyk) [reply src/main/java/com/bookshop/service/impl/UserServiceImpl.java]: there is only existsById so I left it as it is

File: src/main/java/com/bookshop/config/SecurityConfig.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+                .addFilterBefore(jwtAuthenticationFilter,
+                        UsernamePasswordAuthenticationFilter.class)
+                .userDetailsService(userDetailsService)
+                .build();
+    }
+
+    @Bean
+    public AuthenticationManager authenticationManager(
+            AuthenticationConfiguration authenticationConfiguration
+    ) throws Exception {
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/config/SecurityConfig.java]:
```suggestion
            AuthenticationConfiguration authenticationConfiguration) throws Exception {
```

File: src/main/java/com/bookshop/controller/AuthController.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    @PostMapping("/register")
+    @ResponseStatus(HttpStatus.CREATED)
+    public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto request) {
+        return userService.register(request);
+    }
+
+    @Operation(summary = "Login",
+                description = "Login user by their credentials")
+    @PostMapping("/login")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/controller/AuthController.java]: just FYI

`OK` is the default response status, so you can omit it

remember about this in future :)

File: src/main/java/com/bookshop/security/AuthenticationService.java [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class AuthenticationService {
+    private final JwtUtil jwtUtil;
+    private final AuthenticationManager authenticationManager;
+
+    public UserLoginResponseDto authenticate(UserLoginRequestDto requestDto) {
+        final Authentication authentication = authenticationManager.authenticate(
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/security/AuthenticationService.java]:
```suggestion
        Authentication authentication = authenticationManager.authenticate(
```

File: src/main/java/com/bookshop/security/JwtAuthenticationFilter.java [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    private static final String BEARER = "Bearer ";
+    private final JwtUtil jwtUtil;
+    private final UserDetailsService userDetailsService;
+
+    @Override
+    protected void doFilterInternal(
+            HttpServletRequest request,
+            HttpServletResponse response,
+            FilterChain filterChain
+    ) throws ServletException, IOException {
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/security/JwtAuthenticationFilter.java]:
```suggestion
            FilterChain filterChain) throws ServletException, IOException {
```

File: src/main/java/com/bookshop/security/JwtUtil.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+                .compact();
+    }
+
+    public boolean isValidToken(String token) {
+        try {
+            Jws<Claims> claimsJws = Jwts.parser()
+                    .verifyWith(secret)
+                    .build()
+                    .parseSignedClaims(token);
+
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/security/JwtUtil.java]: redundant blank line

File: src/main/java/com/bookshop/security/CustomUserDetailsService.java [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        return userRepository.findByEmail(username).orElseThrow(
+                () -> new EntityNotFoundException("There is no user with email: " + username));
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/security/CustomUserDetailsService.java]: so `UsernameNotFoundException` or `EntityNotFoundException` ?:)

remove the `throws`

File: src/main/java/com/bookshop/service/UserService.java [hunk 22]
```diff
@@ -0,0 +1,9 @@
+package com.bookshop.service;
+
+import com.bookshop.dto.user.UserRegistrationRequestDto;
+import com.bookshop.dto.user.UserResponseDto;
+import com.bookshop.exception.RegistrationException;
+
+public interface UserService {
+    UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException;
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/service/UserService.java:8]:
```suggestion
    UserResponseDto register(UserRegistrationRequestDto request);
```

File: src/main/java/com/bookshop/service/impl/UserServiceImpl.java (hunk 10 above)
Comment (NazarSmith228) [reply src/main/java/com/bookshop/service/impl/UserServiceImpl.java]: @TetianaDenchyk no, this is wrong :)

you can declare your custom method `existsByEmail`

File: src/main/java/com/bookshop/service/impl/UserServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+    private final RoleRepository roleRepository;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/service/impl/UserServiceImpl.java]:
```suggestion
    public UserResponseDto register(UserRegistrationRequestDto request) {
```

File: src/main/java/com/bookshop/validator/annotation/Isbn.java [hunk 24]
```diff
@@ -1,25 +1,22 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.hibernate.validator.constraints.ConstraintComposition;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = {IsbnValidator.class})
+@ConstraintComposition
+@ReportAsSingleViolation
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/validator/annotation/Isbn.java]: you don't have any additional constraints - therefore you don't need these annotations

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/com/bookshop/controller/BookController.java [hunk 25]
```diff
@@ -33,12 +34,14 @@ public class BookController {
     @GetMapping
     @Operation(summary = "Get all books by page",
             description = "Get a list of all books at x page with y size")
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/controller/BookController.java:37]: shouldn't admin user also have a role `USER`? or you design your app in a way that admin does not have same priviliges as user has?

File: src/main/java/com/bookshop/dto/user/UserLoginRequestDto.java [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+
+public record UserLoginRequestDto(
+        @NotBlank
+        @Email
+        @Size(max = 255)
+        String email,
+        @NotBlank
+        @Size(min = 7, max = 20)
+        String password
+) {
```
Comment (NazarSmith228) [inline src/main/java/com/bookshop/dto/user/UserLoginRequestDto.java:15]:
```suggestion
        String password) {
```

File: src/main/java/com/bookshop/service/UserService.java (hunk 22 above)
Comment (NazarSmith228) [reply src/main/java/com/bookshop/service/UserService.java:8]: @TetianaDenchyk you haven't removed `throws` here :)

**Review by NazarSmith228 — APPROVED** [review]
LGTM, but pls pay attention to my remarks

**Pull request merged**